shaka/src/media/pipeline_monitor.cc

```diff
@@ -69,14 +69,56 @@ void PipelineMonitor::Stop() {
   thread_.join();
 }
 
+bool AtEnd(const BufferedRanges& ranges, double time, double duration) {
+  if (std::isnan(duration)) {
+    return false;
+  }
+
+  const double relative_time = time - duration;
+
+  if (relative_time >= 0) {
+    return true;
+  }
+
+  // With presentation time tracked with the help of audio device/data a
+  // possible scenario is that audio track is effectively shorter than
+  // advertised duration, or implicit duration defined by video track. In such
+  // circumstances playback time is unable to reach duration exactly. Being
+  // |kNearEndRangeThreshold| close to end of playback, we assume end of
+  // presentation on reaching end time of last buffered range (which itself is
+  // an intersection of ranges of tracks). Threshold of |kAssumeEndThreshold|
+  // compensates for possible discrepancy of external clock processing last
+  // packet with respective update of presentation time.
+  //
+  // To be even more precise and ensure exactly last video frame is displayed
+  // from available data, we could possibly notify external clock of expected
+  // duration and allow media time reporting past its data derived from audio
+  // track. In this case audio playback subsystem having no real data but with
+  // known duration would be able to use audio hardware defined clock to
+  // continue advancing time up to duration value and take playback exactly to
+  // duration time.
```

Comment:
I'm a bit confused about what is happening here.

First, the buffered ranges are based on the buffered encoded frames from MSE.  So if the buffered content doesn't reach the end, it may be bad content.  I'd be fine with allowing a small gap at the end for buffered ranges. We'd just need to handle the gap similar to `IsBufferedUntil` at the top.

But this comment seems to suggest this is caused by the audio player not reading to the end of the content.  Does your second paragraph refer to an alternative workaround?  I would think it would be better to tell the audio player about the duration so it can play to the end instead of forcing playback to end before the actual end.  If this is the case, I would suggest waiting until #60 so the `MediaPlayer` can handle changes to the duration and can report it to the audio player.

Or is this a problem with the decoder not providing frames to the end of the content?  It seems odd that you would have content but playback/decoding would stop a half second before the end.  If the problem is with the last frame not having a duration, maybe it would be better to have the decoder just duplicate the duration from the frame before it?